core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -155,19 +154,88 @@ private TableMetadata internalApply() {
+        }
+      }
+    }
+    //At this point idsToRetain includes all snapshots that are within their branches lifecycle AND the global expiration age
+    //but it could still be insufficient for global min snapshots.
+    //We now look in the expiration set to see what are the latest snapshots that should be kept
+    //ToDo: This is most likely over-engineered.
+    //There's probably a simpler way to keep track of what are snapshots that should be retained for the global
+    Set<Snapshot> snapshotsToExpire = base.snapshots()
+        .stream()
+        .filter(snapshotId -> !idsToRetain.contains(snapshotId))
+        .collect(Collectors.toSet());
+    if (idsToRetain.size() < globalMinSnapshots) {
+      long snapshotsRequiredForGlobal = globalMinSnapshots - idsToRetain.size();
+      //Keep a heap of the "snapshotsRequiredForGlobal"-latest snapshots
+      Queue<Snapshot> heap = new PriorityQueue<>(new SnapshotTimestampComparator());
+      for (Snapshot snapshot : snapshotsToExpire) {
+        if (heap.size() < snapshotsRequiredForGlobal || snapshot.timestampMillis() > heap.peek().timestampMillis()) {
+          heap.offer(snapshot);
+        }
+        if (heap.size() > snapshotsRequiredForGlobal) {
+          heap.poll();
+        }
+      }
+      //At this point the heap should have the elements we want to retain.
+      snapshotsToExpire.removeAll(heap);
     }
 
-    TableMetadata updateMeta = base.removeSnapshotsIf(snapshot ->
-        idsToRemove.contains(snapshot.snapshotId()) ||
-        (snapshot.timestampMillis() < expireOlderThan && !idsToRetain.contains(snapshot.snapshotId())));
-    List<Snapshot> updateSnapshots = updateMeta.snapshots();
+    TableMetadata updatedMetadata = base.removeSnapshotsIf(snapshot -> snapshotsToExpire.contains(snapshot.snapshotId()));
+    updatedMetadata = updatedMetadata.replaceRefs(referencesToRetain);
+    List<Snapshot> updateSnapshots = updatedMetadata.snapshots();
     List<Snapshot> baseSnapshots = base.snapshots();
-    return updateSnapshots.size() != baseSnapshots.size() ? updateMeta : base;
+    return updateSnapshots.size() != baseSnapshots.size() ? updatedMetadata : base;
+  }
+
+
+  private Set<Long> ancestorsToRetain(long startingSnapshotId,
+                                      Integer minSnapshotsToKeep,
+                                      Long maxSnapshotAge,
+                                      TableMetadata tableMetadata,
+                                      long currentTime) {
+    Snapshot startingSnapshot = tableMetadata.snapshot(startingSnapshotId);
+    List<Long> ancestors = SnapshotUtil.ancestorIds(startingSnapshot, base::snapshot);
+    int ancestorIdx = 0;
+    Set<Long> ancestorsToRetain = Sets.newHashSet();
+    while (ancestorIdx < ancestors.size()) {
+      Snapshot ancestor = tableMetadata.snapshot(ancestors.get(ancestorIdx));
+      long comparisonAge = maxSnapshotAge == null ? globalExpireOlderThan : Math.max(maxSnapshotAge, globalExpireOlderThan);
```

Comment:
This depends on what we want the global expireOlderThan to be used for (if at all). Currently as implemented, the branch does not really "override" the global policy if the global policy has only allows for older snapshots to be expired.